File: TD2/Typescript/session2/td/heritageMultiple/etatSucc.ts
import {
    ZeroParInt,SuccParInt
}from "./natInductif";

import {
    Nat
} from "./naturels";

import{
    NatParInt
}from "./natParInt";

export abstract class EtatSucc implements Nat {
    constructor(private pred: Nat) { }

    chiffre(i : number) : number{
      return parseInt(this.representation().charAt(this.taille() - 1 - i));;
    }

    taille() : number{
      return this.representation().length;
    }

    val() : number{
      return this.pred.val() - 1;
    }

    estNul() : boolean{
      return false;
    }

    predecesseur() : Nat{
      return this.pred;
    }

    creerSuccesseur(predecesseur: Nat): Nat {
        return new SuccParInt(predecesseur);
    }

    creerNatAvecValeur(valeur: number): Nat {
        if (valeur < 0) {
            throw new Error('* Erreur : valeur négative.');
        }
        if (valeur % 1 !== 0) {
            throw new Error('* Erreur : valeur non entière.');
        }
        return new NatParInt(valeur);
    }

    creerZero() : Nat {
      return new ZeroParInt();
    }

    creerNatAvecRepresentation(repDecimale: string): Nat {
        return this.creerNatAvecValeur(parseInt(repDecimale));
    }

    somme(x: Nat): Nat {
        return this.creerNatAvecValeur(x.val() + this.val());
    }

    zero(): Nat {
        return this.creerZero();
    }

    un(): Nat {
        return this.creerNatAvecValeur(1);
    }

    modulo(x: Nat): Nat {
        return this.creerNatAvecValeur(this.val() % x.val());
    }

    div(x: Nat): Nat {
        return this.creerNatAvecValeur(Math.floor(this.val() / x.val()));
    }

    produit(x: Nat): Nat {
        return this.creerNatAvecValeur(x.val() * this.val());
    }

    representation(): string {
        return this.val().toString();
    }

    estEgal(n: Nat): boolean {
        return n.val() == this.val();
    }

}

File: Java/src/session2/td/heritage/EtatSucc.java
package session2.td.heritage;

import session2.td.naturels.Nat;
import session2.td.naturels.NatInductifInt;

public abstract class EtatSucc implements NatInductifInt  {

	private Nat predecesseur;
	
	public EtatSucc(Nat val) {
		this.predecesseur = val;
	}
	
	public int val() {
		return this.predecesseur.val() +1;
	}
	
	public boolean estNul() {
		return this.val() == 0;
	}
	
	public Nat predecesseur() {
		if(this.estNul())
			throw new UnsupportedOperationException("0 n'a pas de prédécesseur.");
		return this.creerNatAvecValeur(this.val() - 1);
	}
	
	public int chiffre(int i) {
		if(i==0) {
			return this.val()%10;
		} else {
			return this.creerNatAvecValeur(this.val()/10).chiffre(i-1);
		}
	}
	
	public int taille() {
		if(this.val()<10) {
			return 1;
		} else {
			return this.creerNatAvecValeur(this.val()/10).taille();
		}
	}
}
